File: QETag.java
package com.qiniu;

/*
java 最新版本 
https://github.com/qiniu/java-sdk/blob/master/src/main/java/com/qiniu/util/Etag.java
android 最新版本
https://github.com/qiniu/android-sdk/blob/master/library/src/main/java/com/qiniu/android/utils/Etag.java
*/

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import javax.xml.bind.DatatypeConverter;

public class QETag {
	private final int CHUNK_SIZE = 1 << 22;
	private final int BUFFER_SIZE = 1 << 16;

	public String urlSafeBase64Encode(byte[] data) {
		String encodedString = DatatypeConverter.printBase64Binary(data);
		encodedString = encodedString.replace('+', '-').replace('/', '_');
		return encodedString;
	}

	public String calcETag(String fileName) throws IOException,
			NoSuchAlgorithmException {
		String etag = "";
		File file = new File(fileName);
		if (!(file.exists() && file.isFile() && file.canRead())) {
			System.err.println("Error: File not found or not readable");
			return etag;
		}
		FileInputStream inputStream = new FileInputStream(file);
		MessageDigest chunkDataDigest = MessageDigest.getInstance("sha1");
		MessageDigest allDigestDigest = MessageDigest.getInstance("sha1");
		byte[] buffer = new byte[BUFFER_SIZE];
		int multi = CHUNK_SIZE / BUFFER_SIZE;
		int count = 0;
		int bytesRead;
		boolean bigFlag = false;
		while ((bytesRead = inputStream.read(buffer)) > 0) {
			if (++count > multi) {
				bigFlag = true;
				allDigestDigest.update(chunkDataDigest.digest());
				chunkDataDigest.reset();
				count = 0;
			}
			chunkDataDigest.update(buffer, 0, bytesRead);
		}
		byte[] hashData = new byte[21];
		if (bigFlag) {
			hashData[0] = (byte) 0x96;
			System.arraycopy(allDigestDigest.digest(chunkDataDigest.digest()), 0, hashData, 1, 20);
		} else {
			hashData[0] = 0x16;
			System.arraycopy(chunkDataDigest.digest(), 0, hashData, 1, 20);
		}
		etag = urlSafeBase64Encode(hashData);
		inputStream.close();
		return etag;
	}

	public static void main(String[] args) {
		if (args.length != 1) {
			System.out.println("Usage: qetag <filename>");
		} else {
			String fileName = args[0];
			QETag etag = new QETag();
			try {
				System.out.println(etag.calcETag(fileName));
			} catch (NoSuchAlgorithmException ex) {
				System.err.println("Unsupported algorithm:" + ex.getMessage());
			} catch (IOException ex) {
				System.err.println("IO Error:" + ex.getMessage());
			}
		}
	}
}

File: QETag.scala
package main

import java.io.{File, FileInputStream}
import java.io.InputStream
import java.security.MessageDigest

object QEtag {
  private val BLOCK_BITS = 22 // Indicate that the blocksize is 4M
  private val BLOCK_SIZE = 1 << BLOCK_BITS

  def apply(fname: String) = etag(fname)
  def apply(in: InputStream, size: Long) = etag(in, size)

  def etag(fname: String): String = {
    val file = new File(fname)
    val in = new FileInputStream(file)
    try {
      etag(in, file.length)
    } finally {
      if (in != null) {
        in.close()
      }
    }
  }

  def etag(in: InputStream, size: Long): String = {
    val sha1 = MessageDigest.getInstance("SHA-1")
    blockCount(size) match {
      case 1 => {
        val buf = new Array[Byte](size.toInt)
        require(in.read(buf) == size)
        urlSafeBase64Encode(0x16.toByte +: sha1.digest(buf))
      }
      case _ => {
        val buf = new Array[Byte](BLOCK_SIZE)
        val sha1Tmp = MessageDigest.getInstance("SHA-1")
        Stream.continually(in.read(buf))
          .takeWhile(_ != -1)
          .foreach(n => {
            sha1.update(sha1Tmp.digest(buf.take(n)))
          })
        urlSafeBase64Encode(0x96.toByte +: sha1.digest())
      }
    }
  }

  private def blockCount(fsize: Long) = ((fsize + (BLOCK_SIZE - 1)) >> BLOCK_BITS).toInt

  private def urlSafeBase64Encode(data: Array[Byte]) = {
    val b64Encoder = new sun.misc.BASE64Encoder()
    val encodedData = b64Encoder.encode(data)
    val encodedString = new String(encodedData)
    encodedString.replace('+', '-').replace('/', '_')
  }
}

object Main {
  def main(args: Array[String]) {
    if (args.length != 1) {
      System.err.println("Usage: qetag <filename>")
      System.exit(1)
    }
    println(QEtag(args(0)))
  }
}
